[Flaking test] pull-kubernetes-unit - TestDirty fails sometimes

### Which jobs are flaking?

pull-kubernetes-unit sometimes fails with data race.
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/115585/pull-kubernetes-unit/1622986339706212352

```sh
~/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver > go test -c -race ./
~/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver > stress ./apiserver.test -test.run=TestDirty                                       
                                                                                                     
/tmp/go-stress-20230217T225843-2898065946                                                                                                                                                                  
I0217 22:58:43.599838  409272 handler_discovery.go:392] Starting ResourceDiscoveryManager                                                                                                                  
--- FAIL: TestDirty (0.00s)                                                                                                                                                                                
    handler_discovery_test.go:165: service handler never pinged                                      
FAIL                                                                                                                                                                                                       
                                                                                                                                                                                                           
                                                                                                                                                                                                           
ERROR: exit status 1                                                                                                                                                                                       
                                                                                                                                                                                                           
                                                                                                                                                                                                           
/tmp/go-stress-20230217T225843-1296271475                                                                                                                                                                  
I0217 22:58:44.015272  409511 handler_discovery.go:392] Starting ResourceDiscoveryManager                                                                                                                  
==================                                                                                                                                                                                         
WARNING: DATA RACE                                                                                                                                                                                         
Write at 0x00c0003064bf by goroutine 93:                                                                                                                                                                   
  k8s.io/kube-aggregator/pkg/apiserver.TestDirty.func1()                                                                                                                                                   
      /home/gjkim/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go:154 +0x55                                                                            
  net/http.HandlerFunc.ServeHTTP()                                                                                                                                                                         
      /usr/local/go/src/net/http/server.go:2109 +0x4d                                                                                                                                                      
  net/http.(*timeoutHandler).ServeHTTP.func1()                                                                                                                                                             
      /usr/local/go/src/net/http/server.go:3407 +0xe1                                                                                                                                                      
                                                                                                                                                                                                           
Previous read at 0x00c0003064bf by goroutine 86:                                                                                                                                                           
  k8s.io/kube-aggregator/pkg/apiserver.TestDirty()                                                                                                                                                         
      /home/gjkim/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go:164 +0x433                                                                           
  testing.tRunner()                                                                                                                                                                                        
      /usr/local/go/src/testing/testing.go:1446 +0x216                                                                                                                                                     
  testing.(*T).Run.func1()                                                                                                                                                                                 
      /usr/local/go/src/testing/testing.go:1493 +0x47                                                                                                                                                      
                                                                                                                                                                                                           
Goroutine 93 (running) created at:                                                                   
  net/http.(*timeoutHandler).ServeHTTP()                                                                                                                                                                   
      /usr/local/go/src/net/http/server.go:3401 +0x5e5                                                                                                                                                     
  k8s.io/kube-aggregator/pkg/apiserver.(*discoveryManager).fetchFreshDiscoveryForService()                                                                                                                 
      /home/gjkim/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:218 +0xf31
  k8s.io/kube-aggregator/pkg/apiserver.(*discoveryManager).syncAPIService()
      /home/gjkim/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:346 +0x156
  k8s.io/kube-aggregator/pkg/apiserver.(*discoveryManager).Run.func1.1()
      /home/gjkim/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:413 +0x12e
  k8s.io/kube-aggregator/pkg/apiserver.(*discoveryManager).Run.func1()
      /home/gjkim/go/src/k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go:418 +0x38

Goroutine 86 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1493 +0x75d
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.g
```

### Which tests are flaking?

TestDirty

### Since when has it been flaking?

I think the test was added 3 months ago with the flaking state.

### Testgrid link

https://testgrid.k8s.io/presubmits-kubernetes-blocking#pull-kubernetes-unit

### Reason for failure (if possible)

There are two issues.

https://github.com/kubernetes/kubernetes/blob/3165fe19b34359ed65efb692cf231afc4d257176/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go#L133-L167

First, updating and reading a bool-type variable is not thread-safe.

and there is also another problem checking the state before completely syncing it.

### Anything else we need to know?

_No response_

### Relevant SIG(s)

/sig api-machinery

--- a/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
+++ b/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
@@ -23,6 +23,7 @@ import (
 	"sort"
 	"strconv"
 	"strings"
+	"sync"
 	"sync/atomic"
 	"testing"
 	"time"
@@ -38,19 +39,22 @@ import (
 	discoveryendpoint "k8s.io/apiserver/pkg/endpoints/discovery/aggregated"
 	scheme "k8s.io/client-go/kubernetes/scheme"
 	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
 	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
 )
 
 func newDiscoveryManager(rm discoveryendpoint.ResourceManager) *discoveryManager {
-	return NewDiscoveryManager(rm).(*discoveryManager)
+	dm := NewDiscoveryManager(rm).(*discoveryManager)
+	dm.dirtyAPIServiceQueue = newCompleterWorkqueue(dm.dirtyAPIServiceQueue)
+
+	return dm
 }
 
-// Returns true if the queue of services to sync empty this means everything has
-// been reconciled and placed into merged document
-func waitForEmptyQueue(stopCh <-chan struct{}, dm *discoveryManager) bool {
+// Returns true if the queue of services to sync is complete which means
+// everything has been reconciled and placed into merged document
+func waitForQueueComplete(stopCh <-chan struct{}, dm *discoveryManager) bool {
 	return cache.WaitForCacheSync(stopCh, func() bool {
-		// Once items have successfully synced they are removed from queue.
-		return dm.dirtyAPIServiceQueue.Len() == 0
+		return dm.dirtyAPIServiceQueue.(*completerWorkqueue).isComplete()
 	})
 }
 
@@ -104,7 +108,7 @@ func TestBasic(t *testing.T) {
 
 	go aggregatedManager.Run(testCtx.Done())
 
-	require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+	require.True(t, waitForQueueComplete(testCtx.Done(), aggregatedManager))
 
 	response, _, parsed := fetchPath(aggregatedResourceManager, "")
 	if response.StatusCode != 200 {
@@ -160,7 +164,44 @@ func TestDirty(t *testing.T) {
 	defer cancel()
 
 	go aggregatedManager.Run(testCtx.Done())
-	require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+	require.True(t, waitForQueueComplete(testCtx.Done(), aggregatedManager))
+
+	// immediately check for ping, since Run() should block for local services
+	if !pinged.Load() {
+		t.Errorf("service handler never pinged")
+	}
+}
+
+// Shows that waitForQueueComplete also waits for syncing to
+// complete by artificially making the sync handler take a long time
+func TestWaitForSync(t *testing.T) {
+	pinged := atomic.Bool{}
+	service := discoveryendpoint.NewResourceManager()
+	aggregatedResourceManager := discoveryendpoint.NewResourceManager()
+
+	aggregatedManager := newDiscoveryManager(aggregatedResourceManager)
+
+	aggregatedManager.AddAPIService(&apiregistrationv1.APIService{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "v1.stable.example.com",
+		},
+		Spec: apiregistrationv1.APIServiceSpec{
+			Group:   "stable.example.com",
+			Version: "v1",
+			Service: &apiregistrationv1.ServiceReference{
+				Name: "test-service",
+			},
+		},
+	}, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(3 * time.Second)
+		pinged.Store(true)
+		service.ServeHTTP(w, r)
+	}))
+	testCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go aggregatedManager.Run(testCtx.Done())
+	require.True(t, waitForQueueComplete(testCtx.Done(), aggregatedManager))
 
 	// immediately check for ping, since Run() should block for local services
 	if !pinged.Load() {
@@ -212,7 +253,7 @@ func TestRemoveAPIService(t *testing.T) {
 		aggregatedManager.RemoveAPIService(s.Name)
 	}
 
-	require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+	require.True(t, waitForQueueComplete(testCtx.Done(), aggregatedManager))
 
 	response, _, parsed := fetchPath(aggyService, "")
 	if response.StatusCode != 200 {
@@ -356,7 +397,7 @@ func TestLegacyFallbackNoCache(t *testing.T) {
 	defer cancel()
 
 	go aggregatedManager.Run(testCtx.Done())
-	require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+	require.True(t, waitForQueueComplete(testCtx.Done(), aggregatedManager))
 
 	// At this point external services have synced. Check if discovery document
 	// includes the legacy resources
@@ -465,7 +506,7 @@ func TestLegacyFallback(t *testing.T) {
 	defer cancel()
 
 	go aggregatedManager.Run(testCtx.Done())
-	require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+	require.True(t, waitForQueueComplete(testCtx.Done(), aggregatedManager))
 
 	// At this point external services have synced. Check if discovery document
 	// includes the legacy resources
@@ -534,10 +575,10 @@ func TestNotModified(t *testing.T) {
 
 	// Important to wait here to ensure we prime the cache with the initial list
 	// of documents in order to exercise 304 Not Modified
-	require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+	require.True(t, waitForQueueComplete(testCtx.Done(), aggregatedManager))
 
-	// Now add all groups. We excluded one group before so that AllServicesSynced
-	// could include it in this round. Now, if AllServicesSynced ever returns
+	// Now add all groups. We excluded one group before so that waitForQueueIsComplete
+	// could include it in this round. Now, if waitForQueueIsComplete ever returns
 	// true, it must have synced all the pre-existing groups before, which would
 	// return 304 Not Modified
 	for _, s := range apiServices {
@@ -545,7 +586,7 @@ func TestNotModified(t *testing.T) {
 	}
 
 	// This would wait the full timeout on 1.26.0.
-	require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
+	require.True(t, waitForQueueComplete(testCtx.Done(), aggregatedManager))
 }
 
 // copied from staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler_test.go
@@ -609,4 +650,49 @@ func fetchPath(handler http.Handler, etag string) (*http.Response, []byte, *apid
 	}
 
 	return w.Result(), bytes, decoded
+}
+
+// completerWorkqueue is a workqueue.RateLimitingInterface that implements
+// isComplete
+type completerWorkqueue struct {
+	lock sync.Mutex
+	workqueue.RateLimitingInterface
+	processing map[interface{}]struct{}
+}
+
+var _ = workqueue.RateLimitingInterface(&completerWorkqueue{})
+
+func newCompleterWorkqueue(wq workqueue.RateLimitingInterface) *completerWorkqueue {
+	return &completerWorkqueue{
+		RateLimitingInterface: wq,
+		processing:            make(map[interface{}]struct{}),
+	}
+}
+
+func (q *completerWorkqueue) Add(item interface{}) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	q.processing[item] = struct{}{}
+	q.RateLimitingInterface.Add(item)
+}
+
+func (q *completerWorkqueue) AddAfter(item interface{}, duration time.Duration) {
+	q.Add(item)
+}
+
+func (q *completerWorkqueue) AddRateLimited(item interface{}) {
+	q.Add(item)
+}
+
+func (q *completerWorkqueue) Done(item interface{}) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	delete(q.processing, item)
+	q.RateLimitingInterface.Done(item)
+}
+
+func (q *completerWorkqueue) isComplete() bool {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	return q.Len() == 0 && len(q.processing) == 0
 }

--- a/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
+++ b/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery_test.go
@@ -23,6 +23,7 @@ import (
 	"sort"
 	"strconv"
 	"strings"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -134,7 +135,7 @@ func checkAPIGroups(t *testing.T, api apidiscoveryv2beta1.APIGroupDiscoveryList,
 // Test that a handler associated with an APIService gets pinged after the
 // APIService has been marked as dirty
 func TestDirty(t *testing.T) {
-	pinged := false
+	var pinged atomic.Bool
 	service := discoveryendpoint.NewResourceManager()
 	aggregatedResourceManager := discoveryendpoint.NewResourceManager()
 
@@ -152,7 +153,7 @@ func TestDirty(t *testing.T) {
 			},
 		},
 	}, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		pinged = true
+		pinged.Store(true)
 		service.ServeHTTP(w, r)
 	}))
 	testCtx, cancel := context.WithCancel(context.Background())
@@ -162,7 +163,7 @@ func TestDirty(t *testing.T) {
 	require.True(t, waitForEmptyQueue(testCtx.Done(), aggregatedManager))
 
 	// immediately check for ping, since Run() should block for local services
-	if !pinged {
+	if !pinged.Load() {
 		t.Errorf("service handler never pinged")
 	}
 }

